adobe/react-spectrum
# Fix RSP DropZone's filled message announcement on different platforms

**TLDR:**

Specifically for Chrome + VO, during a virtual drag, VO will not announce any of the labels. However, Chrome + VO is happy to announce labels during a virtual drag so long as there isn't both an `aria-label` and `aria-labelledby` on the element that is receiving focus. As a result, the solution is to add a `div` to the `<VisuallyHidden>` component that receives the`aria-label` prop and is pointed to by the `aria-labelledby`. We can combine all the different elements that make up drop zone’s accessible name and pass it to useLabels as a single `aria-labelledby`.

_____ 

**The Longer Version:** 

Originally, the issue was that some platforms were not announcing the filled message that says “Drop File to Replace”. 

Upon further investigation, it seemed that there was a larger issue, specifically for Chrome + VO. In short, during a virtual drag, it wasn’t announcing any labels and would skip to the `aria-describedby`. At least for the other platforms, if it wasn’t announcing the filled message, it was at least announcing the labels. 
Funny enough, Chrome + VO has no problem announcing the labels when you were not in a virtual drag and just focusing on the visually hidden button. This problem is specific to virtual dragging. 

We haven’t pinpointed what the exact issue is, especially since this seems to be specific to Chrome + VO. However, we do have a workaround that seems acceptable enough. It hinges on the fact that Chrome + VO is capable of announcing labels during a virtual drag BUT ONLY under a specific condition which is that the visually hidden button only have either an `aria-label` OR an `aria-labelledby` but not together. 

The solution is to add a `div` to the `<VisuallyHidden>` component that recieves the`aria-label` prop and is pointed to by the `aria-labelledby`. We can combine all the different elements that make up drop zone’s accessible name and pass it to useLabels as a single `aria-labelledby`. Chrome + VO is happy to announce this. 

I will mention that Chrome + VO is still rather finicky and doesn’t always announce the filled message (though it will at least announce the other labels). Also, depending on how quickly you decide to drag and drop, it might just skip announcing the label entirely. I recommend when testing to go slowly and to give the screenreader some time to talk.

Things that we tried while debugging:
1. Creating a `<Draggable />` and `<DropTarget />` component that used the hooks to see if the same issue still applied. The answer is that it did. Seems to be an issue with drag and drop and not specifically `<DropZone />`. Here’s the [codesandbox](https://codesandbox.io/s/drag-and-drop-investigation-7j7nkl)
2. Created a minimal reproduction with no drag and drop to see if there was an issue with combining aria-labels and aria-labelledby’s on the same element. Here’s the [codesandbox](https://codesandbox.io/s/more-aria-label-dropzone-testing-spy9j8)
3. Clear out the aria-describedby that is passed by dropButtonProps by setting it to null. Reads out only ‘button’ on a virtual drag. 
4. Commented out the announcements from DragManager. Labels were still not being announced. 

## ✅ Pull Request Checklist:


## 📝 Test Instructions:

When the dropzone is filled, the label "Drop file to replace" will always be announced.

1. Turn on screenreader
2. Go to "WithDraggable"
3. Tab to focus on the visually hidden button. Should say "DropZone Drag and drop here"
4. Enter a virtual drag, when dragging over the DropZone the first time, it should say "DropZone Drag and drop here". Drag over the DropZone a second time, it should say "DropZone Drag and drop here Drop file to replace"
5. When the dropzone is in a filled state, focusing on the visually hidden button should announce "DropZone Drag and drop here Drop file to replace"
7. Go to "DropZone Example With Draggable Object" in RAC section of Storybook
8. Tab to focus on the visually hidden button. Should say "DropZone DropZone Area"
9. Enter a virtual drag, when dragging over the DropZone, it should say "DropZone DropZone Area"


## 🧢 Your Project:
